plutus-core/plutus-core/src/PlutusCore/Parser/Type.hs

```diff
@@ -58,15 +70,57 @@ pType = choice $ map try
     , varType
     ]
 
-defaultUniType :: Parser (SomeTypeIn DefaultUni)
-defaultUniType = choice $ map try
-  [ inParens defaultUniType
-  , SomeTypeIn DefaultUniInteger <$ symbol "integer"
-  , SomeTypeIn DefaultUniByteString <$ symbol "bytestring"
-  , SomeTypeIn DefaultUniString <$ symbol "string"
-  , SomeTypeIn DefaultUniUnit <$ symbol "unit"
-  , SomeTypeIn DefaultUniBool <$ symbol "bool"
-  ] --TODO complete all defaultUni types
+-- | Parser for built-in type applications.
+defaultUniApplication :: Parser (SomeTypeIn (Kinded DefaultUni))
+defaultUniApplication = do
+    -- Parse the head of the application.
+    f <- defaultUni
+    -- Parse the arguments.
+    as <- many defaultUni
+    -- Iteratively apply the head to the arguments checking that the kinds match and
+    -- failing otherwise.
+    foldM tryUniApply f as
+
+-- | Parser for built-in types (the ones from 'DefaultUni' specifically).
+--
+-- 'Kinded' is needed for checking that a type function can be applied to its argument.
+-- I.e. we do Plutus kind checking of builtin type applications during parsing, which is
+-- unfortunate, but there's no way we could construct a 'DefaultUni' otherwise.
+--
+-- In case of kind error no sensible message is shown, only an overly general one:
+--
+-- >>> :set -XTypeApplications
+-- >>> :set -XOverloadedStrings
+-- >>> import PlutusCore.Error
+-- >>> import PlutusCore.Quote
+-- >>> let runP = either show display . runQuoteT . parseGen @ParserErrorBundle defaultUni
+-- >>> runP "(list integer)"
+-- "(list integer)"
+-- >>> runP "(bool integer)"
+-- "ParseErrorB (ParseErrorBundle {bundleErrors = TrivialError 13 Nothing (fromList [Tokens ('(' :| \"\"),Tokens ('b' :| \"ool\"),Tokens ('b' :| \"ytestring\"),Tokens ('d' :| \"ata\"),Tokens ('i' :| \"nteger\"),Tokens ('l' :| \"ist\"),Tokens ('p' :| \"air\"),Tokens ('s' :| \"tring\"),Tokens ('u' :| \"nit\")]) :| [], bundlePosState = PosState {pstateInput = \"(bool integer)\", pstateOffset = 0, pstateSourcePos = SourcePos {sourceName = \"test\", sourceLine = Pos 1, sourceColumn = Pos 1}, pstateTabWidth = Pos 8, pstateLinePrefix = \"\"}})"
+--
+-- This is to be fixed.
+--
+-- One thing we could do to avoid doing kind checking during parsing is to parse into
+--
+--     data TextualUni a where
+--         TextualUni :: TextualUni (Esc (Tree Text))
+--
+-- i.e. parse into @Tree Text@ and do the kind checking afterwards, but given that we'll still need
+-- to do the kind checking of builtins regardless (even for UPLC), we don't win much by deferring
+-- doing it.
+defaultUni :: Parser (SomeTypeIn (Kinded DefaultUni))
+defaultUni = choice $ map try
+    [ inParens defaultUniApplication
+    , someType @_ @Integer <$ symbol "integer"
```

Comment:
Not sure how to handle type applications this way. Plus we can't derive `Enum`, because of the type applications. Plus we'd also need to `Enum` for `Kinded DefaultUni` rather than just `DefaultUni`, which is probably even harder. And we don't add new built-ins types too often unlike built-in functions, so I think we can afford having a bit of duplication here, given that fixing it seems quite non-trivial.